micrometer-core/src/main/java/io/micrometer/core/instrument/binder/httpcomponents/PoolingHttpClientConnectionManagerMetricsBinder.java

```diff
@@ -0,0 +1,222 @@
+            .description(METER_ROUTE_CONNECTIONS_DESC)
+            .tags(tags).tag(TAG_CONNECTIONS_STATE, "leased")
+            .register(registry);
+        poolRoutePendingGauge = MultiGauge.builder(METER_ROUTE_PENDING)
+            .description(METER_ROUTE_PENDING_DESC)
+            .tags(tags)
+            .register(registry);
+    }
+
+    /**
+     * Updates the list of routes accessed by the monitored http client.
+     *
+     * Call this method periodically if you want to monitor thread pools for
+     * every accessed route (target scheme, host and port). For example:
+     *
+     * <pre>
+     *     &#x40;Bean
+     *     public RouteUpdater routeUpdater(List&#60;PoolingHttpClientConnectionManagerMetricsBinder&#62; metricsBinders) {
+     *         return new RouteUpdater(metricsBinders);
+     *     }
+     *
+     *     private static class RouteUpdater {
+     *         private final HashSet&#60;PoolingHttpClientConnectionManagerMetricsBinder&#62; metricsBinders = new HashSet&#60;&#62;();
+     *
+     *         private RouteUpdater(Collection&#60;PoolingHttpClientConnectionManagerMetricsBinder&#62; metricsBinders) {
+     *             this.metricsBinders.addAll(metricsBinders);
+     *         }
+     *
+     *         &#x40;Scheduled(fixedRate = 10000)
+     *         public void updateRoutes() {
+     *             metricsBinders.forEach(PoolingHttpClientConnectionManagerMetricsBinder::updateRoutes);
+     *         }
+     *     }
+     * </pre>
+     *
+     * BE CAREFUL! Call this if and only if you are sure that your http client
+     * accesses only a limited number of remote schemes, hosts and ports. DO NOT
+     * USE THIS METHOD if your HttpClient accesses a big or unlimited number of
+     * routes (i.e. if the target host depends on the user input).
+     */
+    @SuppressWarnings("WeakerAccess")
+    public void updateRoutes() {
+        Set<HttpRoute> routes = connectionManager.getRoutes();
+        poolRouteMaxGauge.register(routesToRows(routes, PoolStats::getMax));
+        poolRouteAvailableGauge.register(routesToRows(routes, PoolStats::getAvailable));
+        poolRouteLeasedGauge.register(routesToRows(routes, PoolStats::getLeased));
+        poolRoutePendingGauge.register(routesToRows(routes, PoolStats::getPending));
+    }
+
+    private Iterable<MultiGauge.Row<?>> routesToRows(Set<HttpRoute> routes, Function<PoolStats, Integer> valueFunction) {
+        return routes.stream()
+            .map((route) -> routeToRow(route, () -> valueFunction.apply(connectionManager.getStats(route))))
+            .collect(Collectors.toSet());
+    }
+
+    private MultiGauge.Row<Supplier<Number>> routeToRow(HttpRoute route, Supplier<Number> valueFunction) {
+        Tags tags = Tags.of(
+            "target.host", route.getTargetHost().getHostName(),
+            "target.port", String.valueOf(route.getTargetHost().getPort()),
+            "target.scheme", route.getTargetHost().getSchemeName()
```

Comment:
When I tried to build a dashboard, I saw that separating the route values into three different tags might cause some difficulties. For example, let's say your client accesses the following routes:

 * `http`, `foo.example.com`, `80`
 * `http`, `bar.example.com`, `80`
 * `https`, `foo.example.com`, `443`

You might now want to create a board, which shows statistics for all three routes, meaning three combinations of the values `http`/`https` + `foo.example.com`/`bar.example.com` + `80`/`443`. But that's where I'm stuck now. Combining all tags, would also create the route `https`, `bar.example.com`, `80`, which is never really called by the application.

The initial idea of putting these values into separate tags was, to be able to create statistics for all `http` calls, or statistics for all calls to a specific host. But the more I think about it, I guess that in most cases this is not really needed, and that it would be enough to put these values just as a single `route` tag, for example:

 * `http://foo.example.com:80`
 * `http://bar.example.com:80`
 * `https://foo.example.com:443`

What do you think? Do you have an idea, how to query the original solution in Prometheus? How is it with other monitoring backends?

Reply:
I don't really have the resulting metrics at hand, but what about:
```
sum(httpcomponents_httpclient_pool_route_connections) by (target.scheme)
... by (target.host,target.port)
```

Reply:
I want to create a Grafana board, containing a *repeating graph panel*, so one panel for each route. I can only iterate through one of the `target_*` tags, but not through all.

Do you have a solution for this?

Reply:
You could create a composite tag:

* in Micrometer via a metrics filter, 
* in Prometheus as a relabeling step on collection
* in Prometheus via a recording rule